### ISSUE TITLE
ASP.NET Core Runtime 3.1.0 installer contains not signed assemblies

### ISSUE THREAD

**Issue by gaborposz** [dotnet/aspnetcore#18091]
Hello,
We are using Device Guard on machines where we install our product, and it prohibits the execution of not-signed assemblies. Unfortunately the [ASP.NET Core Runtime 3.1.0 Hosting Bundle installer](https://dotnet.microsoft.com/download/dotnet-core/thank-you/runtime-aspnetcore-3.1.0-windows-hosting-bundle-installer) (as well as other .NET Core installers) contains not signed assemblies, thus it cannot be installed with enabled Device Guard:

> Code Integrity determined that a process (\Device\HarddiskVolume1\Windows\Temp\{71947B89-AF78-4F6D-930C-F82ADDE5881B}\.cr\dotnet-hosting-3.1.0-win.exe) attempted to load \Device\HarddiskVolume1\Windows\Temp\{D71E2EDC-B0DF-4210-9930-C48080496705}\.ba\wixstdba.dll that did not meet the Enterprise signing level requirements or violated code integrity policy.

As far as I know the ASP.NET Core Runtime installer is based on the WIX Toolset, and the problematic assembly (_wixstdba.dll_) is also coming from there, but it was already found and fixed in WIX, see [Issue 5329](https://github.com/wixtoolset/issues/issues/5329).

Can you upgrade to a new WIX version where this problem is already fixed, or can you find another way to make the installer 100% signed?

We would primarily need the ASP.NET Core Runtime 3.1.0 Hosting Bundle fixed, but fixing it in other .NET Core installers is also appreciated.

Regards,
Gabor

**Comment by jkotalik** [comment]
@joeloff

**Comment by dougbu** [comment]
I can't tell if the WiX fix is in the current stable release (3.11.2). If yes, that would be a simple upgrade. Otherwise, I'd be concerned about moving to a prerelease WiX version (4.0.0.5512-pre at the moment).

Anyone know whether the fix made it into 3.11.2?

**Comment by gaborposz** [comment]
I'm not sure either, but there were some changes committed to 3.11.2 regarding "signing":
[https://github.com/wixtoolset/wix3/commits/master](https://github.com/wixtoolset/wix3/commits/master)

It think it worth a try, a patch version upgrade shouldn't hurt anyway :)

**Comment by joeloff** [comment]
We don't sign the WiX DLLs that are linked into the bootstrapper. wixstdba is the native bootstrapper that forms the bottom layer of the bundle and is a standard DLL that is part of Wix. Typically only the engine and the bundle is signed. We'd have to maintain a private copy of WiX that we sign with 3rd party certificates. That's certainly possible.

Looking at the changes in the PR, that appears to simplify the process of signing additional payload items that can be added to a bundle.

I'd advise against flipping to 4.x since there are a lot of changes which may require additional authoring changes too depending on how much the authoring schema's have changed.

**Comment by gaborposz** [comment]
Hello,
Any update on this issue?

I've seen there was also some discussion in the [original WiX issue](https://github.com/wixtoolset/issues/issues/5329).

As I understand 
- WiX 3.11.2 will not solve it, 
- WiX 4.x is not an option, 
- and _"maintaining a private copy of WiX that was signed with 3rd party certificates"_ is also difficult. 

How do you see, is it still possible to fix it, and if yes, then how long would it take?

I'm asking because if you cannot fix it in an upcoming .NET Core patch, then we will need to look for a workaround soon :(

Regards,
Gabor

**Comment by joeloff** [comment]
Just fixing the hosting bundle won't solve the problem. The hosting bundle ships the runtime bundle, so once that starts to install I'd expect you to hit a similar issue.

While some of the WiX DLLs can be signed, typically custom action DLLs are linked into a compiler extensions, and then extracted and embedded into MSIs. There's no way to sign those without rebuilding wix from source and signing those files.

The underlying issue is a bit more complex. Here's what WiX typically does

1. You produce your DLL and sign it, e.g. wixstdba
2. You then produce the necessary setup authoring and compile / link that along with the DLL into a wixlib.
3. The wixlib often gets linked into a second DLL, e.g. a compiler extension

When the bundle is produced, the underlying files will be extracted. While we can sign some static files like those in step #1, step 2 would require rebuilding all the extensions from source to ensure they are consuming signed binaries.

The same applies for custom actions, especially the default ones that are at the core of WiX toolset and the only way to be 100% sure is to rebuild WiX from source and sign everything, so we effectively maintain our own copy.

@gaborposz you will run into this issue with any WiX based installer. How are you handling that?

**Comment by gaborposz** [comment]
I asked a co-worker (who is responsible about our installer) how we deal with this problem, and it turned out that we do pretty much that you proposed:
We "forked" WiX several years ago, and built a custom installer from it, and it is signed by our own certificate.

I keep looking for your results with WiX 3.11.2.
I guess -if it helps- then the Runtime bundle's installer should be upgraded too, right?

**Comment by dougbu** [comment]
We're now using WiX version `3.14.0-dotnet` in dotnet/aspnetcore's master branch but are still using `3.11.1` in release/3.1. Should we be updating the version there to address this problem❔ That is, is the problem resolved in more recent WiX versions❔

**Comment by gaborposz** [comment]
Hello,
We are still struggling with this problem every time when we update to a new .NET Core 3.1.x patch version: It costs about 2-3 PD for each new versions to create a DeviceGuard catalog file and sign it with our own certificate, so that we can execute the ASP.NET Core Runtime installer with enabled DeviceGuard from our own installer.

So yes, a solution in the .NET Core 3.1 LTS version would be highly appreciated.

However, I don't know if the newer WiX version (3.11.2 or newer) will really solve it (based on @joeloff's previous answer hopefully yes), and I don't know either if the Runtime bundle's installer has been updated meanwhile or not. Who could help in that one?

**Comment by joeloff** [comment]
@dougbu 3.14 was mostly to address the arm64 support, but if that was a fork we created then those files could potentially be signed by us.

**Comment by joeloff** [comment]
The 3.14.0-dotnet. We could technically sign the content of 3.11.x as well. We simply need to grab the binaries.zip, sign the content and then host that zip internally and ensure that everyone builds using the internal zip toolset. I might be missing something, but we should give it a try.

**Comment by wtgodbe** [comment]
Fixed by https://github.com/dotnet/aspnetcore/pull/29879

**Comment by gaborposz** [comment]
Hello,
I've tested the 3.1.13 installer, and unfortunately the problem is not solved at all :(

I still get the same error message if I try to install .NET Core 3.1.13 on a machine with enabled Device Guard:

![DeviceGuardError](https://user-images.githubusercontent.com/47426565/111758060-4fcc5900-889c-11eb-8d32-367f534babf3.png)

**Comment by wtgodbe** [comment]
Hmm, when I inspect the 3.1.13 hosting bundle locally, the `wixstdba.dll` within is signed (whereas 3.1.12 wasn't):

<img width="300" alt="wixsig1" src="https://user-images.githubusercontent.com/14283640/112534313-244be180-8d68-11eb-90dc-348cea03aa5e.PNG">
<img width="297" alt="wixSig2" src="https://user-images.githubusercontent.com/14283640/112534325-29a92c00-8d68-11eb-9d70-630af3fb458d.PNG">

Not sure what would still be causing the issue - are there any other logs you can attach? @joeloff any other ideas?

**Comment by joeloff** [comment]
Is the machine trying to upgrade an old 3.1 runtime? The upgrade will cause the old EXE to run to uninstall itself, which would then extract its payload which is likely still the unsigned copy.

**Comment by wtgodbe** [comment]
@gaborposz were you able to try out installing 3.1.13 fresh? We believe that if you manually uninstall earlier 3.1 runtimes first, running the 3.1.13 hosting bundle should work without error.

**Comment by gaborposz** [comment]
@wtgodbe: I will check this, but only after Easter ;)
I'll try to give you feedback before the end of the next week.

**Comment by gaborposz** [comment]
I rechecked, and there was no previous .NET Core version installed on the machine, it was a fresh install.

However, I extracted the _wixstdba.dll_ from the temporal install directory, and checked it's signature:
It's signed indeed, but by "_WiX Toolset (.NET Foundation)_".

![wixstdba_signature](https://user-images.githubusercontent.com/47426565/113711297-13df2380-96e5-11eb-9e6c-b7dc0bdbf0d2.png)

The .NET Core installer itself was signed by "_Microsoft Corporation_", and that was accepted by Device Guard, but it seems "_WiX Toolset (.NET Foundation)_" is not accepted by Device Guard as an trusted published.

Any idea how to proceed?

**Comment by joeloff** [comment]
You likely need to add the root certificate to your certificate store.

![image](https://user-images.githubusercontent.com/7409981/113740208-5b4dbb80-96b5-11eb-8bcd-ed43b47f7cc4.png)

That way, if anything else uses the DigiCert root and it gets revoked for some reason, validation would fail for anything that uses the root cert,

From the Start menu, launch certlm.msc and see whether DigiCert is a trusted root

![image](https://user-images.githubusercontent.com/7409981/113740611-b67fae00-96b5-11eb-99fb-1c77bd7703d4.png)

**Comment by gaborposz** [comment]
I've checked the certificate chain, and it was fine, the root certificate was also trusted.

However, I talked to a college (our Device Guard expert), and it turned out that our Device Guard policy does not include every certificate that is trusted on OS level, but we need to add accepted certificates one by one.

So thanks for the fix and the support, the problem is already on our side, hence I will close the issue.

Regards,
Gabor
